TMEDIA-852 : Create Picture Component

ecarlisle opened the pull request:
## Ticket

- [TMEDIA-852 : TMEDIA-852 : Create Picture Component](https://arcpublishing.atlassian.net/browse/TMEDIA-852)

## Description

A `Picture` component and `Picture.Source` subcomponent was added to render `<picture>` and `<source>` tags.

## Acceptance Criteria

- Use `npm run generate:component` to scaffold out “Picture” component in the arc-themes-components repo
     - Type “picture” into the response for the component name 
- Replace placeholder jsx with a `<picture>` element 
     - Picture element should set component class name and any additional classnames. 
- Create storybook stories using a fallback <img> element within the <picture> element, using  <Image>'s component
```
<Picture 
  // other picture sub components
  <Image 
    src="my-image"
    alt="Description of the image"
    height={1600}
    width={900}
    resizerOptions={{
      auth: "abc123",
      filter: 70,
      smart: true,
    }}
  />
/>
```
```
<picture>
  <!-- other source tags -->
  <img
    src="my-image?auth=abc123&width=1600&height=900&filter=70&smart=true"
    width="1600"
    height="900"
    <!-- other unrelated properties -->
  >
</picture>
```
- Create `Picture.Source` sub combonent. The sub component will be setup like the carousel
     - The source should follow a pattern similar to the carousel component where the source element is a sub component. (See carousel reference of [child](https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.00/src/components/carousel/_children/Item.jsx), [mapping](https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.00/src/components/carousel/index.jsx#L409), and[ parent example](https://61eeeb27215b5f003acdf18f-otyyhrauqw.chromatic.com/?path=/docs/components-carousel--carousel).)
- Takes in an object resizerOptions. Note: A source image token is only compatible with one source image
```
<Picture>
  <Picture.Source 
    src="my-image-portrait"
    media="(min-width: 800px)"
    height={1600}
    width={900}
    resizerOptions={
      auth: "xyz123",
      filter: 70,
      smart: true,
    }
  />
   <Picture.Source 
    src="my-image-skinnier-screen"
    media="(max-width: 799px)"
    height={50}
    width={100}
    resizerOptions={{
      auth: "fff1111",
      filter: 70,
      smart: true,
    }}
  />
  
  <Image
    src="my-image"
    height={1600}
    width={900}
    resizerOptions={{
      auth: "abc123",
      filter: 70,
      smart: true,
    }}
    alt="Description of the Image"
  />
</Picture>
```
```
<picture>
  <source 
    srcset="my-image-portrait?auth=xyz123&width=900&height=1600&filter=70&smart=true" 
    media="(min-width: 800px)"
    width="1600"
    height="900"
  >
  <source 
    srcset="my-image-skinnier-screen?auth=fff1111&width=100&height=50&filter=70&smart=true" 
    media="(max-width: 799px)"
    width="100"
    height="50"
  >
  <img 
    src="my-image?auth=abc123&width=1600&height=900&filter=70&smart=true"
    width="1600"
    height="900"
    alt="Description of the Image"
  >
</picture>
```
- If no media query is available from `mediaCondition`, do not render the source tag
- If no src, do not render the `source` tag 
- It’s ok if no height or width provided. Do not render those properties if missing.
- Storybook examples should show various options for media
     - media queries min and max width
     - Not rendering the source tag if no `mediaCondition`
- Image source mapping functions should be handled in the utils folder
- Image should not be able to be stretched past its intrinsic size
- Test coverage should be 100% 
- Documentation should be written in the style of other components, including usage and proptypes. The audience should be for developers as well as some basic information for non-technical users. 
     - Include links to the resizer docs https://github.com/WPMedia/arc-photo-resizer/tree/main/docs/usage#query-parameters
-  Note: “types” file like “webp” are not necessary using v2 resizer

## Test Steps

1. Checkout branch - `git checkout {TMEDIA-852-create-picture-component}`
2. Update dependencies - `npm i`
3. Run Storybook `npm run storybook`
4. Check out the new storybook documentation

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [X] Confirmed all the test steps a reviewer will follow above are working.
- [X] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [X] Confirmed relevant documentation has been updated/added.
- [X] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**

ecarlisle committed:
Add picture component and picture.cource subcomponent.

ecarlisle committed:
Add picture to main index file.

ecarlisle committed:
Add picture to main scss file.

ecarlisle committed:
Change order of picture component in main index file.

ecarlisle committed:
Move format-srcset to /src/utils/

ecarlisle committed:
Complete Storybook documentation.

ecarlisle committed:
Update inline comment and param documnentation.

ecarlisle committed:
Update Source params and add param defaults.

ecarlisle committed:
Change required on Source src param.

nschubach commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,45 @@
+
+const Source = ({ height, media, resizerOptions, resizerURL, src, width, ...rest }) => {
+	if (media && src) {
+		if (resizerOptions && resizerOptions.auth) {
```
```suggestion
		if (resizerOptions?.auth) {
```

ecarlisle replied:
KIS! 👍🏼

JackHowa commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,145 @@
+		resizerURL="https://resizer.com/v2/"
+		resizedOptions={{
+			auth: "secret",
+		}}
```
```suggestion
		}}
		/>
```

component jsx should be closed

nschubach replied:
Noted this as well

ecarlisle replied:
Great catch! Thanks!

JackHowa replied:
different problem; has not been addressed https://github.com/WPMedia/arc-themes-components/pull/127#discussion_r936056847

JackHowa replied:
addressed

JackHowa commented on src/components/picture/index.test.jsx:
```diff
@@ -0,0 +1,38 @@
+import Source from "./_children/source";
+
+describe("Picture component", () => {
+	it("render a pciture element with a nested source element", () => {
```
```suggestion
	it("render a picture element with a nested source element", () => {
```

JackHowa commented on src/utils/format-srcset/index.js:
```diff
@@ -0,0 +1,14 @@
+const formatSrcset = (src, resizerOptions, width, height) =>
+	// TODO : Update comment
```
todo comment? is this ready for review?  @ecarlisle

nschubach replied:
Missed this.  Yeah, need to address this.

ecarlisle replied:
@JackHowa , @nschubach :
Thanks, missed this one. Updated. Wanted to make note of the `width` and `height` arguments overriding values that might also be sent in through `resizerOptions`. I didn't resolve this PR issue as I wanted to make sure we're all amenable to the util approach and how it will also work with `Image`. Thanks!

JackHowa replied:
what's your question?

ecarlisle replied:
@JackHowa : The approach used in `formatSrcset` is slightly different from the code in image here:
https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.0.1/src/components/image/index.jsx#L36-L48
I think the defensive coding is a good measure. I'm wanting to confirm we're ok with this approach as it will replace that code in the `Image` component.

nschubach replied:
I vote for making the change in `Image` to use this so they are consistent.

JackHowa replied:
tldr: Overriding seems like more work and upkeep for an unsupported usecase. 

I personally think it's unnecessary logic to have the width and height overriding the option's values: that would be misusing the api based on the docs in Image component. 

But if this is the desire of the team, I don't really think it's a huge deal. Just make sure this optional logic is documented somewhere (what takes precedence over what in what cases). Also that this logic is tested if it's going to stay. 

For example, if a client doesn't want the width and height set on the image component (bad practice for cls on `<img>`), then they can hack us so to speak. They would use the height and width in the resize options -- not including it in the `<img>` or `<source>` itself's height and width. Also, if a customer puts height and width in the resizer object and in the top-level, they might be confused which we pick.

ecarlisle replied:
@JackHowa : Thank you very much for your feedback! 👍🏼  I've integrated the srcset-naming util into the `Image` component. Please let me know if you find anything concerning.

JackHowa replied:
Image and PIcture documentation has not been updated to describe which width and height takes precedence, with documentation in storybook  `Just make sure this optional logic is documented somewhere (what takes precedence over what in what cases).`

JackHowa replied:
height and width are not set on the source element if the height and width of the srcset are generated from the resizedOptions. there should be a test for this as well @ecarlisle

JackHowa replied:
I addressed this in commit

JackHowa commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,145 @@
+
+Crucially, an "auth" key will allow the image to be resized. This object will resize the image to the specified dimensions and other properties given in the query parameters in the `src`.
+
+The `Picture` component can use zero or more `Picture.Source` subcomponents and should also include one `Image` as a supportive fallback. Those are the two types of children a Picture will support.
```
Not sure why someone would use the picture component but not a picture.source. Not sure we'd want to encourage zero
```suggestion
The `Picture` component can take in `Picture.Source` subcomponents and should also include one `Image` as a supportive fallback. Those are the two types of children a Picture will support.
```

nschubach replied:
It's per standard.  A `picture` must have one `img` and zero or more `source`s.

https://html.spec.whatwg.org/multipage/embedded-content.html#the-picture-element

![image](https://user-images.githubusercontent.com/2287238/182227552-02d12929-c751-4574-ac73-f200467bbd47.png)

JackHowa replied:
cool no worries

nschubach commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,45 @@
+			return (
+				<source
+					{...rest}
+					className="c-picture__source"
```
```suggestion
```
There should never be a reason to style the source... it's also not tokenized.

ecarlisle replied:
Good call! I was semantically following example from the [List](https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.0.1/src/components/price/_children/List.jsx#L3) and [Sale](https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.0.1/src/components/price/_children/Sale.jsx#L3) from [Price](https://github.com/WPMedia/arc-themes-components/tree/arc-themes-release-version-2.0.1/src/components/price), but this tag needs no class. I'll remove all as per commit suggestions. Thanks!

nschubach commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,45 @@
+				/>
+			);
+		}
+		return <source {...rest} className="c-picture__source" media={media} srcSet={src} />;
```
```suggestion
		return <source {...rest} media={media} srcSet={src} />;
```
There should never be a reason to style the source... it's also not tokenized.

nschubach commented on src/components/picture/_children/source.test.jsx:
```diff
@@ -0,0 +1,43 @@
+		const { container } = render(<Source src="test.jpeg" media="(max-width: 799px)" />);
+		expect(container.firstChild).toMatchInlineSnapshot(`
+		<source
+		  class="c-picture__source"
```
```suggestion
```
There should never be a reason to style the source... it's also not tokenized.

nschubach commented on src/components/picture/_children/source.test.jsx:
```diff
@@ -0,0 +1,43 @@
+		);
+		expect(container.firstChild).toMatchInlineSnapshot(`
+		<source
+		  class="c-picture__source"
```
```suggestion
```
There should never be a reason to style the source... it's also not tokenized.

nschubach commented on src/components/picture/index.test.jsx:
```diff
@@ -0,0 +1,38 @@
+		  class="c-picture"
+		>
+		  <source
+		    class="c-picture__source"
```
```suggestion
```
There should never be a reason to style the source... it's also not tokenized.

JackHowa commented on src/utils/format-srcset/index.js:
```diff
@@ -0,0 +1,14 @@
+const formatSrcset = (src, resizerOptions, width, height) =>
```
can this also be used in the image component? I know that would be a helpful work for reusability that we were talking about

ecarlisle replied:
@JackHowa : Yes, this util was made to format the Resizer v2 params and dimensional options for the `Image` querystring as per these lines:
https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.0.1/src/components/image/index.jsx#L36-L48

I wasn't certain if changing `Image` to use this utility was scoped in this ticket. That being said, I feel that it's a small-effort, high-reward change and I'd be happy to include it. Please let me know how you'd like me to proceed. Thanks!

JackHowa replied:
yeah good idea! I think adding the util to the image was something we talked about and it's in the ticket: 

> The interface for configuring the <Image> component will match the <Picture.Source>: that is deliberate. They should share logic in some instances in the util folder. 

<img width="768" alt="Screen Shot 2022-08-01 at 17 33 53" src="https://user-images.githubusercontent.com/5950956/182256771-e75447e2-5ec7-4f64-bef4-5816ab34274d.png">

ecarlisle replied:
@JackHowa : Thank you. I have integrated the utility with image. The changes are ready for review.

JackHowa replied:
I addressed.

nschubach commented on src/components/picture/index.jsx:
```diff
@@ -0,0 +1,26 @@
+	/** Class name(s) that get appended to default class name of the component */
+	className: PropTypes.string,
+	/** The text, images or any node that will be displayed within the component - Only Picture.Source and Image children items are allowed */
+	children: PropTypes.node.isRequired,
```
Although it's not a requirement for this story, and I'm not requiring it... we probably should look into validating this further.  `<picture />` elements should have at least one `<img />` by standard.

ecarlisle replied:
Agreed. I don't believe that `<picture><source/></picture>` will render without including an `<img/>`.
cc: @JackHowa

ecarlisle replied:
As per our discussion, we can not make this a requirement at this time.

nschubach commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,145 @@
+		resizedOptions={{
+			auth: "secret",
+		}}
+	<Image
```
```suggestion
	/>
	<Image
```
Missing closing tag

nschubach commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,145 @@
+		width={200}
+		height={200}
+		resizerURL="https://resizer.com/v2/"
+		alt="A cat meowing"
```
Should this have the resizer options as well, err... `resizedOptions` (auth is required, no?)

ecarlisle replied:
Good call! Thanks!

nschubach reviewed (changes requested):
Some minor things I think.  source classes, and documentation

ecarlisle committed:
Remove unneeded className on Source tag.

Co-authored-by: nschubach <nschubach@gmail.com>

ecarlisle committed:
Remove unneeded className on Source tag.

Co-authored-by: nschubach <nschubach@gmail.com>

ecarlisle committed:
Remove unneeded className on Source tag.

Co-authored-by: nschubach <nschubach@gmail.com>

ecarlisle committed:
Update src/components/picture/_children/source.test.jsx

Remove unneeded className on Source tag.

Co-authored-by: nschubach <nschubach@gmail.com>

ecarlisle commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,145 @@
+
+The `Picture` component and `Picture.Source` subcomponent render the HTML5 `<picture>` and `<source>` tags, respectively. Similar to the Image, the Picture component displays images as both a means of content and content support. Picture, however, allows for different images (with respective resize options) to be used explicitely for different media queries. This allows for different Art Directio ...
+
+For `resizedOptions`, here are the query parameters that are available:
```
@JackHowa : Wanted to ask about this. This is incorrect as this param on `Picture` is currently `resizerOptions` as per the acceptance criteria on the ticket. 
> Takes in an object resizerOptions. Note: A source image token is only compatible with one source image

This is arguably confusing though as `Image` uses a [resizedOptions](https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.0.1/src/components/image/index.jsx#L112-L113) param for the same object. I think they should both use the same param name. I like `resizerOptions` and think it's a more intuitive name but I can change it to `resizedOptions` to be consistent w/ `Image`.
cc: @nschubach

JackHowa replied:
must've been a typo on my end. Use the same prop name from the Image component in the Picture component

JackHowa replied:
this has not been addressed. `resizedOptions` should be used in all places. we want to ensure uniformity and reusability between the image component and picture components

@ecarlisle

JackHowa replied:
I've unresolved the comment after it was resolved by the pr creator

ecarlisle replied:
@JackHowa : 
Thank you! This naming convention was being used in the `format-image-resizer-src` util. I've changed this. `resizedOptions` is being used in all places.

JackHowa replied:
Updated

JackHowa replied:
addressed

ecarlisle committed:
Update src/components/picture/index.stories.mdx

Correct syntax on Storybook example code.

Co-authored-by: Jack Howard <jack.howard@washpost.com>

ecarlisle committed:
Add missing closing tag.

Co-authored-by: nschubach <nschubach@gmail.com>

ecarlisle committed:
Correct spelling error on test description.

Co-authored-by: Jack Howard <jack.howard@washpost.com>

ecarlisle committed:
Remove unneeded className on Source tag.

Co-authored-by: nschubach <nschubach@gmail.com>

ecarlisle committed:
Update comment in resizerOptions util.

ecarlisle commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,44 @@
+	/** The URL of the resizer service. Should have a trailing slash */
+	resizerURL: PropTypes.string,
+	/** The URL to an image to load and display. Should not have a leading slash */
+	src: PropTypes.string,
```
@JackHowa : Wanted to ask about this. In the ticket, we have the following acceptance criteria.
> If no src, do not render the source tag

On line 6 I am setting a conditional for `(media && src)` and returning null if that's not the case. To be consistent with `Image`, should I instead make `src` a required field and only check for `(media)` in the conditional?

cc: @nschubach

nschubach replied:
I'd make `media` and `src` required.  Then you wouldn't need to check for either of them.  Thinking though, we should probably properly test the thrown errors:

https://jestjs.io/docs/expect#tothrowerror

matthewroach replied:
If media and src are required - we shouldn't test the prop types throwing an error as that's testing the integration of anything library

ecarlisle replied:
@matthewroach : Thank you. Do you agree that making the props required to stay consistent w/ `Image` would be preferable to making the `if (media && src) {` check?

matthewroach replied:
We should lean on prop types for required checks over using conditional logic within the block if we are just going to short circuit and exit on missing props like you have here

nschubach replied:
Still need the conditional though to fulfill the AC, so I had him add that condition back in.  The `PropTypes` `required` only warns but still renders unfortunately.

ecarlisle replied:
@matthewroach : Sounds great! I do have a check to not render `Source` if no `src` or `media`. That's in the ticket's acceptance criteria. 👍🏼

matthewroach commented on src/components/picture/index.jsx:
```diff
@@ -0,0 +1,26 @@
+	</picture>
+);
+
+Picture.Source = Source;
```
For consistency we could map the `Image` to be a child component like 
```jsx
Picture.Image = Image;
```

ecarlisle replied:
Sounds good! I've updated Source and the Storybook documentation to account for the change.

matthewroach commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,147 @@
+import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
+
+import Picture from ".";
+import Source from "./_children/source";
```
This shouldn't be needed as you should never import the children components

ecarlisle replied:
Thanks! Removed.

matthewroach replied:
I see this marked as resolved - but the code is still here

matthewroach commented on src/components/picture/index.test.jsx:
```diff
@@ -0,0 +1,37 @@
+import { render, screen } from "@testing-library/react";
+
+import Picture from ".";
+import Source from "./_children/source";
```
You should never consume the children components directly. This should be consumed via `Picture.Source`

ecarlisle replied:
Hi @matthewroach,
Wanted to confirm this as we seem to be doing this with `Price`. 
https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.0.1/src/components/price/index.jsx#L3-L4
The only difference is that with `Price`, the files in the `./_children` folder are init-capped.
Please let me know if I'm misunderstanding. Thanks!

matthewroach replied:
The comment here relates to the test file, you are referencing a component file with you comment

ecarlisle replied:
Yes, sorry, this response was meant for this comment. 
https://github.com/WPMedia/arc-themes-components/pull/127/files#r935474353

ecarlisle replied:
And sorry, that comment is for another request. Thanks! This has been removed.

matthewroach commented on src/components/picture/index.test.jsx:
```diff
@@ -0,0 +1,37 @@
+	it("render a picture element with a nested source element", () => {
+		const { container } = render(
+			<Picture>
+				<Source src="test.jpeg" media="(max-width: 799px)" />
```
You should never consume the children components directly. This should be consumed via Picture.Source

ecarlisle replied:
Thanks! Updated.

matthewroach commented on src/components/picture/index.test.jsx:
```diff
@@ -0,0 +1,37 @@
+		const ADDITIONAL_CLASSES = "additionalClass1 additionalClass2";
+		render(
+			<Picture className={ADDITIONAL_CLASSES} data-testid="test-picture">
+				<Source src="test.jpeg" media="(max-width: 799px)" />
```
You should never consume the children components directly. This should be consumed via Picture.Source

ecarlisle replied:
Thanks! Updated.

matthewroach commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,147 @@
+
+# Picture
+
+The `Picture` component and `Picture.Source` subcomponent render the HTML5 `<picture>` and `<source>` tags, respectively. Similar to the Image, the Picture component displays images as both a means of content and content support. Picture, however, allows for different images (with respective resize options) to be used explicitely for different media queries. This allows for different Art Directio ...
```
```suggestion
The `Picture` component and `Picture.Source` subcomponent render the HTML5 `<picture>` and `<source>` tags, respectively. Similar to the Image, the Picture component displays images as both a means of content and content support. Picture, however, allows for different images (with respective resize options) to be used explicitly for different media queries. This allows for different Art Direction approaches for specific viewport dimensions.
```

ecarlisle committed:
Use optional chaining to shorten condition.

Co-authored-by: nschubach <nschubach@gmail.com>

ecarlisle committed:
Rename resizerOptions param to resizedOptions.

ecarlisle committed:
Correct spelling error in Storybook documentation.

Co-authored-by: Matthew Roach <matthewroach@users.noreply.github.com>

ecarlisle committed:
Change props on Source:

- Make `src` and `media` required props.
- Remove check for `src` and `media` from business logic.

ecarlisle committed:
Remove unneeded tests since making  and  required props in .

ecarlisle committed:
Set imported  to be a subcomponent of .

ecarlisle committed:
Update documentation to refer to  as .

ecarlisle committed:
Merge branch 'TMEDIA-852-create-picture-component' of github.com:WPMedia/arc-themes-components into TMEDIA-852-create-picture-component

ecarlisle committed:
Correct test to get coverage for auth key checking.

ecarlisle committed:
Ensure Source subcomponents are being called with Picture.Source.

ecarlisle committed:
Remove unneeded import from test.

matthewroach commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,41 @@
+			/>
+		);
+	}
+	return <source {...rest} media={media} srcSet={src} />;
```
Do we need `width` and `height` attributes here?

ecarlisle replied:
@matthewroach :Thank you, great catch! Added both height and width.

matthewroach commented on src/components/picture/_children/source.test.jsx:
```diff
@@ -0,0 +1,33 @@
+describe("Source subcomponent", () => {
+	it("render unresized image an auth key has not been provided", () => {
+		const { container } = render(
+			<Source src="test.jpeg" media="(max-width: 799px)" resizedOptions={{}} />
```
```suggestion
			<Source src="test.jpeg" media="(max-width: 799px)" />
```

ecarlisle replied:
@matthewroach : Thank you! There were three tests using these empty objects. I removed these.

matthewroach commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,156 @@
+
+import Picture from ".";
+import Source from "./_children/source";
+import Image from "../image";
```
Should not be importing the `Image` component here

ecarlisle replied:
@matthewroach : Thank you! Fixed.

ecarlisle committed:
Update Storybook

- remove unneeded imports.
- update Picture to correct subcomponent name.

ecarlisle committed:
Rename image resizer src utility name.

ecarlisle committed:
Re-add condition for having src and media props.

ecarlisle committed:
Re-add tests for checking media and src props.

ecarlisle committed:
Change image to use format-image-resizer-src util.

ecarlisle committed:
Exclude unneeded resizedOptions={{}} from test cases.

ecarlisle committed:
Add height and width attributes on unresized source.

matthewroach commented on src/components/picture/index.stories.mdx:
```diff
@@ -0,0 +1,154 @@
+				auth: "secret",
+			}}
+		/>
+		/>
```
Too many closing tags

ecarlisle replied:
Thank you! Fixed.

JackHowa reviewed (changes requested):
Documentation has not been updated to reflect the logic that you can set height and width in the resizer in two different way. Please see previous comment https://github.com/WPMedia/arc-themes-components/pull/127#discussion_r935894442

JackHowa commented:
Please use `resizedOptions` instead of `resizerOptions` in all places based on previous feedback. I'm seeing resizerOptions in some places. 

@ecarlisle

ecarlisle committed:
Remove redundant vlosing tag on Storybook example code.'

ecarlisle committed:
Correct  to  on  util.

ecarlisle commented:
> Please use `resizedOptions` instead of `resizerOptions` in all places based on previous feedback. I'm seeing resizerOptions in some places.
> 
> @ecarlisle

@JackHowa : Thank you! There were a few instances of that in the `format-image-resizer-src` util. I've corrected that. `resizedOptions` is now used in all places.

ecarlisle committed:
Add Storybook documentation regarding dimension handling for Image and Picture components.

ecarlisle committed:
Adjust punctuation on documentation.

ecarlisle commented:
> Documentation has not been updated to reflect the logic that you can set height and width in the resizer in two different way. Please see previous comment [#127 (comment)](https://github.com/WPMedia/arc-themes-components/pull/127#discussion_r935894442)

Thank you, @JackHowa !
I have added the following to the Storybook documentation for both Image and Picture:
> Image dimensions may be provided as `width`, `height` props or `resizedOptions` parameters. If dimensions are provided in both manners the `width`, `height` props will be used, overriding the `resizedOptions` parameters.

matthewroach commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,42 @@
+
+const Source = ({ height, media, resizedOptions, resizerURL, src, width, ...rest }) => {
+	if (!media || !src) return null;
+	if (resizedOptions?.auth) {
+		return (
+			<source
+				{...rest}
+				height={height}
+				media={media}
+				srcSet={formatSrcset(resizerURL.concat(src), resizedOptions, width, height)}
+				width={width}
+			/>
+		);
+	}
+	return <source {...rest} media={media} srcSet={src} height={height} width={width} />;
```
The only thing that changes is what is the value of `srcSet` - what you think about the following? 

```suggestion
	return (
		<source
			{...rest}
			height={height}
			media={media}
			srcSet={
				resizedOptions?.auth
					? formatSrcset(resizerURL.concat(src), resizedOptions, width, height)
					: src
			}
			width={width}
		/>
	);
```

nschubach replied:
This works for me... but I like fewer `return`s.

ecarlisle replied:
Much cleaner! Thank you!

matthewroach commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,42 @@
+Source.defaultProps = {
+	resizedOptions: {},
+	resizerURL: "",
+	width: null,
+	height: null,
```
Find this a little contradictory, the properties have to be `number` as noted in the propTypes but we set defaults to `null` - we can just remove the defaults for `width` and `height`

nschubach replied:
I'm fine with removing these nulls, it would just be undefined which would be falsy and work with the logic anyway.

ecarlisle replied:
I'm fine as well for the same reason. Should we remove them from `Image` as well?
https://github.com/WPMedia/arc-themes-components/blob/arc-themes-release-version-2.0.1/src/components/image/index.jsx#L99-L100

ecarlisle replied:
I've updated `source.jsx` to remove default props for `width` and `height`.

matthewroach commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,42 @@
+import formatSrcset from "../../../utils/format-image-resizer-src";
+
+const Source = ({ height, media, resizedOptions, resizerURL, src, width, ...rest }) => {
+	if (!media || !src) return null;
```
Shouldn't this be an `&&` as both props are marked with `isRequired`

```suggestion
	if (!media && !src) return null;
```

nschubach replied:
If neither of them are there, it will render null, which is appropriate.

`If not media or not src`, render null if either of them are missing.

`If not media and not src` would mean it renders if one of them is there.

nschubach replied:
The alternative is to wrap the whole return with `if (media && src)` which is the De Morgan boolean inversion of the original.

ecarlisle replied:
I updated to use the De Morgan boolean inversion.

nschubach commented on src/utils/format-image-resizer-src/index.test.js:
```diff
@@ -0,0 +1,57 @@
+				filter: 70,
+				smart: true,
+				width: 100,
+				height: 100,
+			})
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=100&height=100");
+	});
+
+	it("with width and height as explicit params, not as resizer options", () => {
+		expect(
+			formatSrcset(
+				"https://www.test.org/image.jpg",
+				{
+					auth: "xyz123",
+					filter: 70,
+					smart: true,
+				},
+				50,
+				50
+			)
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
+	});
+
+	it("with width and height explicit params overriding dimensions provided as resizer options", () => {
+		expect(
+			formatSrcset(
+				"https://www.test.org/image.jpg",
+				{
+					auth: "xyz123",
+					filter: 70,
+					smart: true,
+					width: 100,
+					height: 100,
+				},
+				50,
+				50
+			)
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
+	});
+});
```
Just a rename to be consistent.

```suggestion
import formatSrc from ".";

describe("format sourceset", () => {
	it("without width or height params", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true");
	});

	it("with width and height included as resizer options", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
				width: 100,
				height: 100,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=100&height=100");
	});

	it("with width and height as explicit params, not as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});

	it("with width and height explicit params overriding dimensions provided as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
					width: 100,
					height: 100,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});
});
```

nschubach replied:
Just a rename of `formatSrcSet` to be consistent.

```suggestion
import formatSrc from ".";

describe("format sourceset", () => {
	it("without width or height params", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true");
	});

	it("with width and height included as resizer options", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
				width: 100,
				height: 100,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=100&height=100");
	});

	it("with width and height as explicit params, not as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});

	it("with width and height explicit params overriding dimensions provided as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
					width: 100,
					height: 100,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});
});
```

nschubach replied:
Just a rename of `formatSrcSet` to be consistent.

```suggestion
import formatSrc from ".";

describe("format sourceset", () => {
	it("without width or height params", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true");
	});

	it("with width and height included as resizer options", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
				width: 100,
				height: 100,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=100&height=100");
	});

	it("with width and height as explicit params, not as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});

	it("with width and height explicit params overriding dimensions provided as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
					width: 100,
					height: 100,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});
});
```

nschubach commented on src/utils/format-image-resizer-src/index.test.js:
```diff
@@ -0,0 +1,57 @@
+				filter: 70,
+				smart: true,
+				width: 100,
+				height: 100,
+			})
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=100&height=100");
+	});
+
+	it("with width and height as explicit params, not as resizer options", () => {
+		expect(
+			formatSrcset(
+				"https://www.test.org/image.jpg",
+				{
+					auth: "xyz123",
+					filter: 70,
+					smart: true,
+				},
+				50,
+				50
+			)
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
+	});
+
+	it("with width and height explicit params overriding dimensions provided as resizer options", () => {
+		expect(
+			formatSrcset(
+				"https://www.test.org/image.jpg",
+				{
+					auth: "xyz123",
+					filter: 70,
+					smart: true,
+					width: 100,
+					height: 100,
+				},
+				50,
+				50
+			)
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
+	});
+});
```
Just a rename of `formatSrcSet` to be consistent.

```suggestion
import formatSrc from ".";

describe("format sourceset", () => {
	it("without width or height params", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true");
	});

	it("with width and height included as resizer options", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
				width: 100,
				height: 100,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=100&height=100");
	});

	it("with width and height as explicit params, not as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});

	it("with width and height explicit params overriding dimensions provided as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
					width: 100,
					height: 100,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});
});
```

ecarlisle replied:
Good catch! Thanks!

nschubach commented on src/utils/format-image-resizer-src/index.test.js:
```diff
@@ -0,0 +1,57 @@
+				filter: 70,
+				smart: true,
+				width: 100,
+				height: 100,
+			})
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=100&height=100");
+	});
+
+	it("with width and height as explicit params, not as resizer options", () => {
+		expect(
+			formatSrcset(
+				"https://www.test.org/image.jpg",
+				{
+					auth: "xyz123",
+					filter: 70,
+					smart: true,
+				},
+				50,
+				50
+			)
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
+	});
+
+	it("with width and height explicit params overriding dimensions provided as resizer options", () => {
+		expect(
+			formatSrcset(
+				"https://www.test.org/image.jpg",
+				{
+					auth: "xyz123",
+					filter: 70,
+					smart: true,
+					width: 100,
+					height: 100,
+				},
+				50,
+				50
+			)
+		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
+	});
+});
```
Just a rename of `formatSrcSet` to be consistent.

```suggestion
import formatSrc from ".";

describe("format sourceset", () => {
	it("without width or height params", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true");
	});

	it("with width and height included as resizer options", () => {
		expect(
			formatSrc("https://www.test.org/image.jpg", {
				auth: "xyz123",
				filter: 70,
				smart: true,
				width: 100,
				height: 100,
			})
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=100&height=100");
	});

	it("with width and height as explicit params, not as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});

	it("with width and height explicit params overriding dimensions provided as resizer options", () => {
		expect(
			formatSrc(
				"https://www.test.org/image.jpg",
				{
					auth: "xyz123",
					filter: 70,
					smart: true,
					width: 100,
					height: 100,
				},
				50,
				50
			)
		).toBe("https://www.test.org/image.jpg?auth=xyz123&filter=70&smart=true&width=50&height=50");
	});
});
```

ecarlisle replied:
I dig consistency!

nschubach commented on src/utils/format-image-resizer-src/index.test.js:
```diff
@@ -0,0 +1,57 @@
+import formatSrcset from ".";
```
Should rename `formatSrcset` to `formatSrc` here for consistency.

ecarlisle replied:
Got this one in your suggested commit. Thanks!

nschubach commented on src/components/picture/_children/source.jsx:
```diff
@@ -0,0 +1,42 @@
+import PropTypes from "prop-types";
+import formatSrcset from "../../../utils/format-image-resizer-src";
```
Should rename `formatSrcset` to `formatSrc` here for consistency.

ecarlisle replied:
Thanks! Updated!

ecarlisle committed:
Invert Source conditional for  and  requirements.

ecarlisle committed:
Use updated util object name `formatSrcset` instead of `formatSrc`.

Co-authored-by: nschubach <nschubach@gmail.com>

ecarlisle committed:
Update name of import util to .

ecarlisle committed:
Remove width and height as being null default props in Source.

JackHowa reviewed (dismissed):
Height and width are not set on the source from resizedOptions https://github.com/WPMedia/arc-themes-components/pull/127#discussion_r936930849

JackHowa committed:
feat: Use only the width and height top-level options. Remove the height and width from the resized options.

JackHowa committed:
docs: Update the disclaimer essentially on the image element to discourage use of the resized options for width, height

JackHowa committed:
docs: Clarify the usage of only the width and height of the source component.

JackHowa committed:
test: Update the width and height of the source element behavior

JackHowa committed:
fix: Update defaults for number of width and height

JackHowa reviewed (commented):
My changes on the resizedOptions has been applied. Other people can review as well

JackHowa committed:
docs: Update the comment to match the expect in the jest test case for resized options

ecarlisle committed:
Resolve merge conflict w/ base branch.

nschubach commented on src/components/image/index.jsx:
```diff
@@ -57,11 +46,13 @@ const Image = ({
+					resizerURL.concat(src),
+					resizedOptions,
+					responsiveImageWidth,
+					aspectRatio !== 0 && height ? responsiveImageWidth / aspectRatio : undefined
```
```suggestion
					aspectRatio !== 0 && height ? Math.floor(responsiveImageWidth / aspectRatio) : undefined
```
per Ryan, we should either floor or round the pixel values here...

ecarlisle committed:
Round calculated height of image using `Math.floor`.

Co-authored-by: nschubach <nschubach@gmail.com>

matthewroach commented on src/components/image/index.jsx:
```diff
@@ -57,11 +46,13 @@ const Image = ({
+					resizerURL.concat(src),
+					resizedOptions,
+					responsiveImageWidth,
+					aspectRatio !== 0 && height ? Math.floor(responsiveImageWidth / aspectRatio) : undefined
```
The `Math.floor` should be put within the `formatSrc` function that way we will not have situations where it's different between components or consumers of the `formatSrc` function

nschubach replied:
Oh right, good point.  Sorry @ecarlisle

nschubach commented on src/utils/format-image-resizer-src/index.js:
```diff
@@ -0,0 +1,25 @@
+			...resizedOptionsNoDimensions,
+			// show height and width in params if available
+			// using optional object key syntax
+			...(width ? { width } : {}),
+			...(height ? { height } : {}),
```
```suggestion
			...(width ? { width: Math.floor(width) } : {}),
			...(height ? { height: Math.floor(height) } : {}),
```

matthewroach replied:
Fixed

matthewroach committed:
Move Math.floor to util

matthewroach committed:
Reduce returns

nschubach reviewed (approved):
Epic

JackHowa commented:
@ecarlisle I'm looking at the storybooks as a client. Let me know if this is helpful for understanding. 

1. it looks like the width and heights of the three pictures in the storybook are all different. is this on purpose? 

![Screen Shot 2022-08-04 at 15 28 01](https://user-images.githubusercontent.com/5950956/182946337-3c4c4218-9eac-4124-8392-642ca5dc1004.png)
![Screen Shot 2022-08-04 at 15 27 51](https://user-images.githubusercontent.com/5950956/182946351-8b0eafda-69ee-4bfc-9cfa-904347c15261.png)
![Screen Shot 2022-08-04 at 15 27 47](https://user-images.githubusercontent.com/5950956/182946360-409c1f84-246c-468b-8322-bd2fc9208479.png)

2. The second story "Picture with different sources" describes using different sources. But I see the same image used across the breakpoints. Am I misunderstanding what the story is showing?

ecarlisle committed:
Change Picture story to show art direction at different viewports.

ecarlisle committed:
Merge branch 'TMEDIA-852-create-picture-component' of github.com:WPMedia/arc-themes-components into TMEDIA-852-create-picture-component

ecarlisle commented:
@JackHowa : Please see updated story. Different crops and dimensions are showed for different viewport widths.

JackHowa reviewed (commented):
Makes more sense visually. It looks like there's one source tag for mobile. Then the default img is being used as a fallback. This is for story `Landscape Picture Source for mobile, portrait Picture Source for desktop`. When we talked about this, I mentioned having various media conditions, like desktop and mobile. 

> Storybook examples should show various options for media
> - media queries min and max width

![Screen Shot 2022-08-05 at 12 23 06](https://user-images.githubusercontent.com/5950956/183129977-66c62519-c2e1-4cbd-952e-04c823b54967.png)

ecarlisle committed:
Correct story showing art direction at different viewport widths.

ecarlisle committed:
Ensure to have media prop on Sources for Art Direction story.

ecarlisle committed:
Remove typo character.

ecarlisle committed:
Adjust story fallback image to image true aspect ratio.

ecarlisle committed:
Adjust story fallback image to have lower quality output.

ecarlisle committed:
Remove typos from Art Direction story.

ecarlisle committed:
Change to Picture component stories:

1. Ensure all uncropped images are resized to appropriate aspect ratio.
2. Remove unneeded <source> from 2nd story.
3. Adjust wording of changed 2nd story.

matthewroach committed:
Add Chromatic viewports to test picture source rendering

ecarlisle committed:
Change Picture Stories

1. The first story is better representred as an `Image` use case and has been removed.
2. The NEW first story shows three image sizes using two `Source`s and one `Image`, shows
   better association with multiple crops used for art direction. An explanation for this
   is provided in the "Docs" mode as the story description.
3. The second story is the `Source` not rendering if `media` attribute is not provided.
4. Main `Picture` storybookbook description has been revised for clarity.``

ecarlisle committed:
Merge branch 'TMEDIA-852-create-picture-component' of github.com:WPMedia/arc-themes-components into TMEDIA-852-create-picture-component
